llvm/lib/Target/SPIRV/SPIRVPreLegalizer.cpp

```diff
@@ -744,79 +744,139 @@ static void insertSpirvDecorations(MachineFunction &MF, MachineIRBuilder MIB) {
-        BasicBlock *CaseBB =
-            Ins[i]->getOperand(1).getBlockAddress()->getBasicBlock();
-        auto It = BB2MBB.find(CaseBB);
-        if (It == BB2MBB.end())
-          report_fatal_error("cannot find a machine basic block by a basic "
-                             "block in a switch statement");
-        MachineBasicBlock *Succ = It->second;
-        ClearAddressTaken.insert(Succ);
-        NewOps.push_back(MachineOperand::CreateMBB(Succ));
-        if (!llvm::is_contained(MBB->successors(), Succ))
-          MBB->addSuccessor(Succ);
-        ToEraseMI.insert(Ins[i]);
-      } else {
-        NewOps.push_back(
-            MachineOperand::CreateCImm(Ins[i]->getOperand(1).getCImm()));
+    for (unsigned i = 0; i < MI->getNumOperands(); ++i) {
+      // The operand is not a register, keep as-is.
+      if (!MI->getOperand(i).isReg()) {
+        NewOps.push_back(MI->getOperand(i));
+        continue;
+      }
+
+      Register Reg = MI->getOperand(i).getReg();
+      MachineInstr *BuildMBB = MRI.getVRegDef(Reg);
+      // The register is not the result of G_BLOCK_ADDR, keep as-is.
+      if (!BuildMBB || BuildMBB->getOpcode() != TargetOpcode::G_BLOCK_ADDR) {
+        NewOps.push_back(MI->getOperand(i));
+        continue;
       }
+
+      assert(BuildMBB && BuildMBB->getOpcode() == TargetOpcode::G_BLOCK_ADDR &&
+             BuildMBB->getOperand(1).isBlockAddress() &&
+             BuildMBB->getOperand(1).getBlockAddress());
+      BasicBlock *BB =
+          BuildMBB->getOperand(1).getBlockAddress()->getBasicBlock();
+      auto It = BB2MBB.find(BB);
+      if (It == BB2MBB.end())
+        report_fatal_error("cannot find a machine basic block by a basic block "
+                           "in a switch statement");
+      MachineBasicBlock *ReferencedBlock = It->second;
+      NewOps.push_back(MachineOperand::CreateMBB(ReferencedBlock));
+
+      ClearAddressTaken.insert(ReferencedBlock);
+      ToEraseMI.insert(BuildMBB);
     }
-    for (unsigned i = MI.getNumOperands() - 1; i > 1; --i)
-      MI.removeOperand(i);
+
+    // Replace the operands.
+    assert(MI->getNumOperands() == NewOps.size());
+    while (MI->getNumOperands() > 0)
+      MI->removeOperand(0);
     for (auto &MO : NewOps)
-      MI.addOperand(MO);
-    if (MachineInstr *Next = MI.getNextNode()) {
+      MI->addOperand(MO);
+
+    if (MachineInstr *Next = MI->getNextNode()) {
       if (isSpvIntrinsic(*Next, Intrinsic::spv_track_constant)) {
         ToEraseMI.insert(Next);
-        Next = MI.getNextNode();
+        Next = MI->getNextNode();
       }
       if (Next && Next->getOpcode() == TargetOpcode::G_BRINDIRECT)
         ToEraseMI.insert(Next);
     }
   }
 
+  // BlockAddress operands were used to keep information between passes,
+  // let's undo the "address taken" status to reflect that Succ doesn't
+  // actually correspond to an IR-level basic block.
+  for (MachineBasicBlock *Succ : ClearAddressTaken)
+    Succ->setAddressTakenIRBlock(nullptr);
```

Comment:
Yes, if we add other instruction taking the address, we need to update this code (one of the reasons I needed to update it was that I added new instructions taking BB addresses).

If we don't remove them, the BB are marked as "address taken". The generic ASM printer will then emit empty basic blocks at the end of the function, with a comment saying this empty BB is emitted because its address was taken by another BB.
I saw a PR to address this as the AsmPrinter level (since we never need this AFAIK), but seems like it's being discussed. So in the meantime, I need to clean-up those BlockAddr.

Reply:
@s-perron @Keenuts This is not related to AsmPrinter and emission of labels, rather this change is a part of a series of PR's improving correctness of emitted MIR between passes for branching instructions and thus increasing the number of passing tests when expensive checks are on. This particular is from https://github.com/llvm/llvm-project/pull/106966 and aims to address this error that we have been getting from MachineVerifier:
https://github.com/llvm/llvm-project/blob/300161761df54f5f85630a8ad0e170d09d119ee3/llvm/lib/CodeGen/MachineVerifier.cpp#L697

Reply:
Ah it's the machine verifier, not the AsmPrinter! thanks.
But yes, it's the labels being generated at the end of the blob with "address taken, but removed by codegen" this additional code is addressing.

Reply:
I guess it doesn't matter much and anyway I'm not 100% sure so please never mind. It's just that, as far as I remember, we addressed your report https://github.com/llvm/llvm-project/pull/87823#issuecomment-2113002300 (resulted from changes in https://github.com/llvm/llvm-project/pull/87823) in the later PR https://github.com/llvm/llvm-project/pull/92390 and it's not about setAddressTakenIRBlock() but rather about https://github.com/llvm/llvm-project/blob/df2a2a072a010d757c23237afc9b7c721702fe2b/llvm/lib/Target/SPIRV/SPIRVPreLegalizer.cpp#L880